[Lee Hyung Woon] iP

lhw-1 opened the pull request:
# Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. 
It's:

* Text-based
* Easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is:

1. Download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```
public class Duke {
    public static void main(String[] args) {
        Bot bot = new Bot();
        bot.start();
    }
}

Commit:
Add Gradle support

lhw-1 committed:
Level-1

lhw-1 committed:
Level-2

lhw-1 committed:
Level-3

lhw-1 committed:
Level-4

lhw-1 committed:
A-TextUiTesting

lhw-1 committed:
temp

lhw-1 committed:
Level-5

lhw-1 committed:
Level-6

lhw-1 committed:
A-Enums

lhw-1 committed:
A-Enums

lhw-1 committed:
Add Level-7 Save feature

lhw-1 committed:
Add Level-8 Dates and Times feature

lhw-1 committed:
Merge branch-Level-7 to master

lhw-1 committed:
Merge branch-Level-8 to master

lhw-1 committed:
Resolve merge conflicts for branch-Level-7 and branch-Level-8

lhw-1 committed:
Add More OOP

lhw-1 committed:
Refactor classes and organize into packages

lhw-1 committed:
Add JUnit Tests

lhw-1 committed:
Create JAR File

lhw-1 committed:
Add JavaDoc comments

lhw-1 committed:
Update Coding Standard

lhw-1 committed:
Add Level-9 Find feature

lhw-1 committed:
Resolve merge conflicts for branch-A-JavaDoc and branch-A-CodingStandard

lhw-1 committed:
Resolve merge conflicts for branch-Level-9

lhw-1 committed:
Update README.md and welcome message

lhw-1 committed:
Update UI display texts

jonas-chow commented on src/main/java/bot/TaskList.java:
```diff
@@ -0,0 +1,89 @@
+ */
+public class TaskList {
+
+    private ArrayList<Task> list;
```
Should the plural form be used? This represents multiple tasks. Perhaps rename it to tasks?

lhw-1 replied:
Good point. Will update - thanks!

jonas-chow commented on src/main/java/bot/UserInterface.java:
```diff
@@ -0,0 +1,121 @@
+        String prefix;
+
+        switch (type) {
+            case 1:
```
There shouldn't be indentation for case clauses

lhw-1 replied:
Yep, I must have missed that part on the guide. Thanks!

jonas-chow commented on src/main/java/bot/Storage.java:
```diff
@@ -0,0 +1,82 @@
+package main.java.bot;
+
+import main.java.task.Task;
```
Consider reordering this import statement (and others like it) to the bottom. Import statement order should be consistent.

lhw-1 replied:
Nice catch - missed that. Thanks!

jonas-chow commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,97 @@
+public class Task {
+
+    protected String task;
+    protected boolean state;
```
Consider renaming this. Booleans should be phrased to sound like booleans. Perhaps isDone might be a better name?

lhw-1 replied:
Good suggestion. Thanks!

jonas-chow reviewed (commented):
Looks good to me! Coding standard mostly followed with only a few minor violations that can be easily ironed out.

jodichoo commented on src/main/java/bot/Parser.java:
```diff
@@ -0,0 +1,66 @@
+
+        }
+    }
+}
```
I like how this method is written so cleanly!

lhw-1 replied:
Thanks! 😄👍

jodichoo commented on src/main/java/bot/Storage.java:
```diff
@@ -0,0 +1,82 @@
+     * @return A TaskList object that will contain all the tasks given to Duke by the user.
+     * @throws DukeException if there is an error loading the local data.
+     */
+    public static TaskList load() throws DukeException {
```
Perhaps you could name this loadTasks to make it clear that a list of tasks is returned?

lhw-1 replied:
Good point - will rename to loadStorage to make it clearer. Thanks for the suggestion!

jodichoo commented on src/main/java/bot/UserInterface.java:
```diff
@@ -0,0 +1,121 @@
+                        : type == 3 ? "(At: " + LocalDateTime.parse(time, inputFormatter).format(outputFormatter) + ")"
+                                : "")
+                + "\n\n" + "You now have " + (length + 1) + (length == 0 ? " task" : " tasks")
+                + " in the list.\n" + "-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~\n");
```
Can consider storing the "-~-~-" divider in a named constant so that the code is easier to understand

lhw-1 replied:
Fair point. Will do!

jodichoo commented on src/main/java/bot/UserInterface.java:
```diff
@@ -0,0 +1,121 @@
+                + taskType + ":\n--> " + prefix + " " + input
+                + (type == 2 ? "(By: " + LocalDateTime.parse(time, inputFormatter).format(outputFormatter) + ")"
+                        : type == 3 ? "(At: " + LocalDateTime.parse(time, inputFormatter).format(outputFormatter) + ")"
+                                : "")
```
This part seems pretty complicated, perhaps you could do this in steps like: 
```java
String taskDesc = taskType + ":\n--> " + prefix + " " + input; 

String taskTime = type == 2
        ? "(By: " + LocalDateTime.parse(time, inputFormatter).format(outputFormatter) + ")"
        : type == 3
            ? "(At: " + LocalDateTime.parse(time, inputFormatter).format(outputFormatter) + ")"
            : "";

System.out.println("-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~\n"
        + "Alright. I've added the following " + taskDesc + taskTime + ...) 
```
Or even place the expression for taskTime into a separate method!

lhw-1 replied:
Noted - made changes. Thanks!

jodichoo commented on src/main/java/command/DoneCommand.java:
```diff
@@ -0,0 +1,49 @@
+                    + "-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~\n");
+
+        } catch (NumberFormatException e) {
+            throw new DukeException("It looks like you did not enter a valid integer for the \"done\" command. Please try again!");
```
This line seems to have exceeded the line length limit (120 chars), perhaps you can add line breaks to improve readability?

lhw-1 replied:
Noted - made changes. Thanks!

jodichoo reviewed (commented):
Hello! Your code is overall pretty clean and readable with good code quality. I didn't see many violations aside from the ones I pointed out 😄

lhw-1 committed:
Merge branch 'add-gradle-support'

lhw-1 committed:
Add CheckStyle detection

lhw-1 committed:
Fix coding standard violations

lhw-1 committed:
Fix CheckStyle violations

lhw-1 committed:
Merge branch 'branch-A-CheckStyle'

lhw-1 committed:
Update .gitignore

lhw-1 committed:
Add initial GUI

lhw-1 committed:
Merge branch 'branch-Level-10'

lhw-1 committed:
Fix UI text display spacing

lhw-1 committed:
Fix JavaDoc and Error Handling for indices

lhw-1 committed:
Update build.gradle with JavaFX dependencies

eugenecsa commented on src/main/java/bot/Bot.java:
```diff
@@ -0,0 +1,175 @@
+    /**
+     * Constructor for the Bot class.
+     *
+     * Takes no parameters.
```
You may want to consider removing this as it is sufficient to just leave the parameters section empty.

lhw-1 replied:
That's a fair point - will remove unnecessary statements for parameters. Thanks!

eugenecsa commented on src/main/java/bot/Bot.java:
```diff
@@ -0,0 +1,175 @@
+        //Scroll down to the end every time dialogContainer's height changes.
+        dialogContainer.heightProperty().addListener((observable) -> scrollPane.setVvalue(1.0));
+    }
+
```
The comments are very helpful and clear in describing the code!

eugenecsa commented on src/main/java/bot/TaskList.java:
```diff
@@ -0,0 +1,89 @@
+     *
+     * @return The size of the TaskList.
+     */
+    public int getSize() { return this.tasks.size(); }
```
Perhaps you can consider putting the method body in the next line to improve readability?

lhw-1 replied:
Oh, it looks like IntelliJ automatically puts the method body in the same line. I'll have to check the settings for this one. Thanks for pointing this out!

eugenecsa commented on src/main/java/bot/UserInterface.java:
```diff
@@ -0,0 +1,167 @@
+
+    private static final DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy - hh:mm a");
+
+    private static final String divider = "-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~\n";
```
As these are constants, you might consider naming them with all uppercase and underscore separation instead?

lhw-1 replied:
That is a very fair point. Will update. Thanks!

eugenecsa commented on src/main/java/bot/UserInterface.java:
```diff
@@ -0,0 +1,167 @@
+                ? " (By: " + LocalDateTime.parse(time, inputFormatter).format(outputFormatter) + ")"
+                : type == 3
+                    ? " (At: " + LocalDateTime.parse(time, inputFormatter).format(outputFormatter) + ")"
+                    : "";
```
Do note that the indentation for wrapped lines should be 8 spaces more than the parent line.

lhw-1 replied:
Thanks for the reminder! I've already extracted some parts of this and simplified the code, so this is no longer an issue.

eugenecsa commented on src/main/java/command/DeleteCommand.java:
```diff
@@ -0,0 +1,48 @@
+     * @param ui The UI of Duke to be invoked by the command.
+     * @return A String representation of Duke's response according to the input given by the user.
+     * @throws DukeException if the input given is not of the correct format.
+     */
```
Perhaps consider adding the `@Override` tag here as the method is from the super class.

lhw-1 replied:
Nice catch! I must have missed that one. Thanks!

eugenecsa commented on src/main/java/task/EventTask.java:
```diff
@@ -0,0 +1,64 @@
+     */
+    @Override
+    public String getTaskState() {
+        return "[E]" + super.getTaskState() + " (At: " + (storedTime.isEmpty() ? time.format(outputFormatter) : storedTime) + ")";
```
Should this line be wrapped? It seems to have exceeded the line length limit.

lhw-1 replied:
That's a good point - I'll check if this has already been addressed, and fix if it has not been. Thanks!

eugenecsa reviewed (commented):
Great work! The code is well organized and easy to read.

I like how there is ample line breaks throughout which improves the readability.

GabrielWLM commented:
Nice work! The code is quite organized and follows coding standards to a large extent!

I agree with the previous reviewers and it seems that you have made your code cleaner than it was. I could not find any more violations. Good job!

lhw-1 committed:
Add assertions to Storage, Command, and Task classes

There are a few classes that make implicit assumptions about the code
behaviour. The save() method in Storage class assumes that the list of
task given has been initialised. The constructors for the Command and
Task classes assume that the input provided is not empty.

Explicitly showing these assumptions via assertions improves the code
readability.

lhw-1 committed:
Improve code quality and enhance code readability

There are a number of classes that have methods comprised of
unnecessary deep-nesting and complicated code logic. The code is
sometimes difficult to follow due to unsuitable variable names.

The task type and stored time in Task classes are displayed using
unnecessarily complicated ternary operations. This pattern is
repeated in the showTaskAdded() method of UserInterface class.

The code logic of execute() methods in TodoCommand, DeadlineCommand,
and EventCommand classes are not obvious and unnecessarily verbose.

Extracting more variables and methods to deal with these cases
significantly improves code quality and enhances code readability.

lhw-1 committed:
Merge pull request #2 from lhw-1/branch-A-CodeQuality

Improve code quality and enhance code readability

lhw-1 committed:
Merge branch 'master' of https://github.com/lhw-1/ip into branch-A-Assertions

lhw-1 committed:
Merge pull request #1 from lhw-1/branch-A-Assertions

Add assertions to Storage, Command, and Task classes

lhw-1 committed:
Fix index issue for done command

lhw-1 committed:
Merge branch 'master' of https://github.com/lhw-1/ip

lhw-1 committed:
Add Extension C-Tagging

lhw-1 committed:
Merge branch 'branch-C-Tagging'

lhw-1 committed:
Improve code quality for JavaDoc, Override methods, and Constant names

lhw-1 committed:
Improve Duke GUI

lhw-1 committed:
Fix Storage Load issue

lhw-1 committed:
Fix tag whitespace issue

lhw-1 committed:
Fix index whitespace issue on done, delete, and find

lhw-1 committed:
Fix index issue on done, delete, and tag

lhw-1 committed:
Add User Guide and Ui.png

lhw-1 committed:
Add User Guide for exit command

lhw-1 committed:
Rename Ui.PNG to Ui.png

lhw-1 committed:
Fix Storage mkdirs() issue

lhw-1 committed:
Fix Find, Done, Delete, Tag empty command parsing issue
